src/context/TodoContext.tsx

```diff
@@ -0,0 +1,139 @@
+import React, { useCallback, useEffect, useState } from 'react';
+import { Todo } from '../types/Todo';
+import { Status } from '../types/Status';
+import { Error } from '../types/ErrorMessage';
+import { getTodos } from '../api/todos';
+import { wait } from '../utils/fetchClient';
+import { getFilteredTodos } from '../utils/getFilteredTodos';
+import { USER_ID } from '../types/userId';
+
+type ContextProps = {
+  todos: Todo[];
+  filteredTodos: Todo[];
+  filter: Status;
+  tempTodo: Todo | null;
+  errorMessage: Error | null;
+  updatingIds: number[];
+  addTodo: (newTodo: Todo) => void;
+  deleteTodo: (todoId: number) => void;
+  updateTodo: (todo: Omit<Todo, 'userId'>) => void;
+  handleSetTempTodo: (newTodo: Todo | null) => void;
+  handleSetErrorMessage: (newError: Error) => void;
+  handleSetUpdatingIds: (id: number | null) => void;
+  filterChange: (newStatus: Status) => void;
+};
+
+export const TodoContext = React.createContext<ContextProps>({
+  todos: [],
+  filteredTodos: [],
+  filter: Status.all,
+  tempTodo: null,
+  errorMessage: Error.none,
+  updatingIds: [],
+  addTodo: () => {},
+  deleteTodo: () => {},
+  updateTodo: () => {},
+  handleSetTempTodo: () => {},
+  handleSetErrorMessage: () => {},
+  handleSetUpdatingIds: () => {},
+  filterChange: () => {},
+});
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const GlobalProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState<Error | null>(null);
+  const [filter, setFilter] = useState(Status.all);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [updatingIds, setUpdatingIds] = useState<number[]>([]);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage(Error.todosLoading);
+
+        return wait(3000).then(() => setErrorMessage(null));
+      });
+  }, []);
+
+  useEffect(() => {
+    wait(3000).then(() => setErrorMessage(Error.none));
+  }, [errorMessage]);
+
+  const filteredTodos = getFilteredTodos(todos, filter);
```

Comment:
since `filter` is available from context it's better to calculate `filteredTodos` in the component you use it (you use it only in one component - `TodoList`). Or at least wrap it into `useMemo`

Reply:
absolutely. I'll move it to the TodoList